### PR TITLE
feat: Add Graph Connector Bot sample app

### PR DESCRIPTION
**Pull request by formulahendry**
[Feature Spec](https://msazure.visualstudio.com/Microsoft%20Teams%20Extensibility/_git/teamsfx-specs?path=/specs/query_custom_data_with_graph_connector.md&_a=preview)

[Demo video](https://microsoft-my.sharepoint-df.com/:v:/p/junjieli/Ec8BITj8_PxOm3U6DeZ81-0BMLhqFRc6OCNfpHHWQO0lcw)

For latest bot query, please below screenshot:
image

Architecture:
image

To reviewers, the overall code is based on https://github.com/OfficeDev/TeamsFx-Samples/tree/dev/bot-sso, so you could mainly focus on the diff part:
* bot code in: [bot/commands/](https://github.com/OfficeDev/TeamsFx-Samples/blob/junhan/connector-bot/graph-connector-bot/bot/commands)
* backend services code in: [bot/services/](https://github.com/OfficeDev/TeamsFx-Samples/blob/junhan/connector-bot/graph-connector-bot/bot/services)

### CODE REVIEW

File: graph-connector-bot/bot/package.json [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    "watch": "nodemon --watch ./src -e ts --exec \"npm run start\""
+  },
+  "repository": {
+    "type": "git",
+    "url": "https://github.com"
+  },
+  "dependencies": {
+    "@azure/msal-node": "^1.1.0",
+    "@microsoft/microsoft-graph-client": "^3.0.1",
+    "@microsoft/teamsfx": "^1.0.0",
```
Comment (adashen) [inline graph-connector-bot/bot/package.json:23]: Please check with Yiqing, there is currently some work undergoing to use 2.0-beta version. If so we may use 2.0-beta

File: graph-connector-bot/bot/commands/query.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  }
+
+  async query(context: TurnContext, text: string, ssoToken: string) {
+    await context.sendActivity("Querying data from Microsoft Graph connectors ...");
+
+    try {
+      const graphQueryService = new GraphQueryService(ssoToken);
+
+      const connectionId = Utils.getConnectionId();
+      const query = text.substring(6);
```
Comment (blackchoey) [inline graph-connector-bot/bot/commands/query.ts:26]: minor: should we use capture group to get the text to query?

Comment (formulahendry) [reply graph-connector-bot/bot/commands/query.ts:26]: As the pattern is simple and extra whitespace does affect the query, so the simple `substring` is enough~

File: graph-connector-bot/bot/index.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    .catch((err) => {
+      // Error message including "412" means it is waiting for user's consent, which is a normal process of SSO, sholdn't throw this error.
+      if (!err.message.includes("412")) {
+        throw err;
+      }
+    });
+});
+
+server.post("/api/notification", async (req, res) => {
+  NotificationHandler.processGraphWebhookChangeNotification(req);
```
Comment (wenytang-ms) [inline graph-connector-bot/bot/index.ts:74]: await ?

File: graph-connector-bot/README.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# Getting Started with Graph Connector Bot Sample
+
```
Comment (MuyangAmigo) [inline graph-connector-bot/README.md:2]: Can we name the sample as `Query custom data ingested via Microsoft Graph connector with Teams command bot`? It's a little bit long but more descriptive, also aligns with what CDA suggested for other sample naming options.

File: graph-connector-bot/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+This sample app showcases how to build custom Graph Connector with Bots in Microsoft Teams and query data using Microsoft Graph Client and TeamsFx SDK.
+
```
Comment (MuyangAmigo) [inline graph-connector-bot/README.md:4]: Suggestion: `This sample ass showcases how to build a Teams command bot that queries custom data ingested into Microsoft Graph using Graph connector.`

File: graph-connector-bot/README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+![Graph Connector Overview](images/graph-connector-bot-demo.gif)
+
+## Prerequisite
+- [NodeJS](https://nodejs.org/en/), fully tested on NodeJS 14, 16
+- An [Azure subscription](https://azure.microsoft.com/en-us/free/)
+- An M365 account. If you do not have M365 account, apply one from [M365 developer program](https://developer.microsoft.com/en-us/microsoft-365/dev-program)
+
```
Comment (MuyangAmigo) [inline graph-connector-bot/README.md:11]: I guess we want to mention that Admin permission is needed in order to test Graph connector

File: graph-connector-bot/README.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+## What you will learn in this sample:
+- How to handle webhook notification when Graph connector is turned on or off in Teams Admin Center.
+- How to use Microsoft Graph API to integrate with Microsoft Graph connector.
+- How to use TeamsFx to build bot to query data from Microsoft Graph connector.
+
```
Comment (MuyangAmigo) [inline graph-connector-bot/README.md:16]: `TeamsFx` -> `TeamsFx SDK`?

File: graph-connector-bot/README.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+### Local Debug the Sample
+1. Clone the repo to your local workspace or directly download the source code.
+1. Download [Visual Studio Code](https://code.visualstudio.com) and install [Teams Toolkit Visual Studio Code Extension](https://aka.ms/teams-toolkit).
+1. Open the project in Visual Studio Code.
+1. Open Debug View (`Ctrl+Shift+D`) and select "Debug (Edge)" or "Debug (Chrome)" in dropdown list.
+1. Press "F5" to open a browser window and then select your package to view sample app. 
+1. Find the Teams package in `build\appPackage\appPackage.local.zip`.
+1. Navigate to [Teams Admin Center](https://admin.teams.microsoft.com/policies/manage-apps), and upload the Teams package.
+    > **Note**: If you are testing in a new debug session again, you need to delete the existing Teams app in Teams Admin Center and upload a new one, since the ngrok endpoint will change when you start a new debug session.
```
Comment (MuyangAmigo) [inline graph-connector-bot/README.md:27]: Can we add where to find the Teams app pacakge?

File: graph-connector-bot/README.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
+# Getting Started with Graph Connector Bot Sample
+
+This sample app showcases how to build custom Graph Connector with Bots in Microsoft Teams and query data using Microsoft Graph Client and TeamsFx SDK.
+
+![Graph Connector Overview](images/graph-connector-bot-demo.gif)
+
+## Prerequisite
+- [NodeJS](https://nodejs.org/en/), fully tested on NodeJS 14, 16
```
Comment (MuyangAmigo) [inline graph-connector-bot/README.md:8]: `Prerequisite` -> `Prerequisites`

File: graph-connector-bot/README.md (hunk 6 above)
Comment (MuyangAmigo) [inline graph-connector-bot/README.md:11]: `M365` -> `Microsoft 365`

File: graph-connector-bot/README.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+> If you don't have an Azure subscription, create [a free account](https://azure.microsoft.com/en-us/free/) before you begin
+1. Open the command palette and select `Teams: Provision in the cloud`. You will be asked to select Azure resource group to provision the sample.
+1. Once provision is completed, open the command palette and select `Teams: Deploy to the cloud`.
+1. Once deployment is completed, you can preview the APP running in Azure. In Visual Studio Code, open `Run and Debug` and select `Launch Remote (Edge)` or `Launch Remote (Chrome)` in the dropdown list and Press `F5` or green arrow button to open a browser.
+1. Use Teams Toolkit to [submit the Teams app to the Teams Admin Center](https://docs.microsoft.com/en-us/microsoftteams/platform/toolkit/publish#publish-to-your-organization).
+1. [Approve the Teams app for publishing to your organization's app store](https://docs.microsoft.com/en-us/microsoftteams/platform/toolkit/publish#admin-approval-for-teams-apps).
+
+## Use the APP in Teams
+1. Navigate to 'Graph Connector' section of the Teams app in [Teams Admin Center](https://admin.teams.microsoft.com/policies/manage-apps), and click 'Grant permissions'.
```
Comment (MuyangAmigo) [inline graph-connector-bot/README.md:38]: `APP` -> `app`

File: graph-connector-bot/bot/index.ts (hunk 3 above)
Comment (formulahendry) [reply graph-connector-bot/bot/index.ts:74]: By design here to not use 'await' in order to immediately return an HTTP 202 and process webhook notification in background.

File: graph-connector-bot/README.md (hunk 8 above)
Comment (formulahendry) [reply graph-connector-bot/README.md:27]: L25: Find the Teams package in `build\appPackage\appPackage.local.zip`

File: graph-connector-bot/bot/index.ts (hunk 3 above)
Comment (formulahendry) [reply graph-connector-bot/bot/index.ts:75]: And thanks for asking~ I have added comment for this line of code.

**Review by blackchoey — APPROVED** [review]

**Pull request merged**